components/layout/model.rs

```diff
@@ -507,61 +522,59 @@ impl ToGfxMatrix for ComputedMatrix {
-    min: Au,
-    max: Option<Au>
+pub struct SizeConstraint {
+    min_size: Au,
+    max_size: Option<Au>,
 }
 
-impl MinMaxConstraint {
-    /// Create a `MinMaxConstraint` for a dimension given the min, max, and content box size for
-    /// an axis
-    pub fn new(content_size: Option<Au>, min: LengthOrPercentage,
-               max: LengthOrPercentageOrNone) -> MinMaxConstraint {
-        let min = match min {
-            LengthOrPercentage::Length(length) => length,
-            LengthOrPercentage::Percentage(percent) => {
-                match content_size {
-                    Some(size) => size.scale_by(percent),
-                    None => Au(0),
-                }
-            },
-            LengthOrPercentage::Calc(calc) => {
-                match content_size {
-                    Some(size) => size.scale_by(calc.percentage()),
-                    None => Au(0),
-                }
+impl SizeConstraint {
+    /// Create a `SizeConstraint` for an axis. The `border` param is present
+    /// when the value of `box-sizing` is `border-box`.
+    pub fn new(container_size: Option<Au>, min_size: LengthOrPercentage,
+               max_size: LengthOrPercentageOrNone, border: Option<Au>)
+               -> SizeConstraint {
+        let min_size = match container_size {
+            Some(container_size) => specified(min_size, container_size),
+            None => if let LengthOrPercentage::Length(length) = min_size {
+                length
+            } else {
+                Au(0)
             }
         };
 
-        let max = match max {
-            LengthOrPercentageOrNone::Length(length) => Some(length),
-            LengthOrPercentageOrNone::Percentage(percent) => {
-                content_size.map(|size| size.scale_by(percent))
-            },
-            LengthOrPercentageOrNone::Calc(calc) => {
-                content_size.map(|size| size.scale_by(calc.percentage()))
+        let max_size = match container_size {
+            Some(container_size) => specified_or_none(max_size, container_size),
+            None => if let LengthOrPercentageOrNone::Length(length) = max_size {
+                Some(length)
+            } else {
+                None
             }
-            LengthOrPercentageOrNone::None => None,
         };
+        // Make sure max size is not smaller than min size.
+        let max_size = max_size.map(|x| max(x, min_size));
 
-        MinMaxConstraint {
-            min: min,
-            max: max
+        if let Some(border) = border {
+            SizeConstraint {
+                min_size: max((min_size - border), Au(0)),
+                max_size: max_size.map(|x| max(x - border, Au(0)))
+            }
+        } else {
+            SizeConstraint {
+                min_size: min_size,
+                max_size: max_size
+            }
         }
     }
 
     /// Clamp the given size by the given `min` and `max` constraints.
+    /// Make sure the size is to be used for content box.
```

Comment:
Can you elaborate on this comment? Technically this can be reused for other purposes.